Docs 7795 update object store tutorial hn (#37)
mulesoft/docs-object-store: modules/ROOT/pages/osv2-tutorial.adoc

. Select the second flow and, in the properties window, change the name to `Retrieve`.
+
When you are done, the flows should look like this:
+
.The screenshot shows (*1*) the Store flow and (*2*) the Retrieve flow.
image::os-tutorial-studio7-flows.png[Store and Retrieve flows in Studio]

. In *Package Explorer* in Studio, right-click `osv2-demo` and select *Anypoint Platform* > *Deploy to CloudHub*.
. At the prompt, select the environment to deploy the app to.
. In the Anypoint Platform *Deploying Application* window, click *Deploy Application*:
+
.The arrow shows the *Deploy Application* button in the *Deploying Application* page.
image::os-tutorial-deploy-app.png[Deploy Application button in the Deploying Application page]
. Click *Open in Browser* to display the app in Runtime Manager.
+
If CloudHub displays a permissions error, select a different environment to deploy to.

You have successfully used Studio 7 to configure an app with an object store.
Next, you test the connection by submitting REST API commands to the Object Store API and using Runtime Manager to view the results.

=== Test Your Studio 7 Application

> After you create the app in Studio 7, you test your app using a `curl` command to submit a REST API command to the Object Store API and use Runtime Manager to view the results.

Reviewer: ```suggestion
After you create your app in Studio 7, you can test it using a `curl` command to submit a REST API command to the Object Store API and use Runtime Manager to view the results.
```

To test your app:

. In *Runtime Manager*, click *Applications* and ensure that the `osv2-demo` app is running.

> . Click the status for the `osv2-demo` app to display its details in the right pane.

Reviewer: How does one "click the status"?

Author: ... the *Status* column?

. Copy the *App url* value:
+
.The arrow shows *App url* in the details pane in Runtime Manager.
image::os-tutorial-rtm-deploy-url.png[App url in Runtime Manager]

> . Run the following `curl` command from a command prompt.

Reviewer: ditto prev. edits to ensure that : directly precedes the example:

```suggestion
. From the command line, run the following `curl` command, replacing _APP-URL_ with the *App url* value that you copied from Runtime Manager:
```

+

> Replace _APP-URL_ with the *App url* value that you copied from Runtime Manager. 

Reviewer: see above

+
[source,console]
----
$ curl -X POST -H "Content-Type: application/json" -d '{ "key": "myKey", "value": "OSv2 Test" }' "http://APP-URL/store";echo
----
+

> The output is an echo:

Reviewer: Will your audience be familiar with this use of "echo"?

+
[source,console]
----
{ "key": "myKey", "value": "OSv2 Test" }
----
+

> . In Runtime Manager, in the details pane, click *Manage application*.

Reviewer: I know I did not edit this earlier, but how about...:

```suggestion
. In the Runtime Manager details pane, click *Manage application*.
```

Author: Not quite correct, but I'll fix.

+

> .The arrow shows *Manage application in the details pane in Runtime Manager.

Reviewer: ```suggestion
.The arrow shows *Manage application* in the details pane in Runtime Manager.
```

image::os-tutorial-rtm-manage-app.png[Manage application in the details pane]
. Click *Object Store* in the left navigation pane.
. Click the object store, the partition, and then the key:
+

> .The screenshot shows (*1*) the *Object Store*, (*2*) the key, and (*3*) the value in the Object store.

Reviewer: prev. sentence uses "object store" rather than "Object store"

+

> image::os-tutorial-rtm-key-in-os.png[Key and value in the Object Store in Runtime Manager]

Reviewer: ditto re. variations of object store....

+
*DO WE NEED THIS DETAIL?*
+
The `myKey` key that you sent from `curl` appears in the Object Store with the value of `[binary value] BINARY`.
The value is binary because Mule 4 wraps values in a Mule object, which causes
them to be visible only as binary in Anypoint Platform.
Behind the scenes, Mule 4 executes binary serialization with the Mule internal serializer.

> The user interface cannot deserialize the object, so it shows that the value is now binary.

Reviewer: ditto prev. edit comments about this text earlier in the set

+
.The arrow shows the *View in Runtime Manager* option in the *...* menu in Design Center.
image::os-tutorial-dc-view-rtm-option.png[View in Runtime Manager menu option]
. In *Runtime Manager*, click *Object Store* in the left navigation pane.
. Click the object store, the partition, and then the key:
+
.The screenshot shows (*1*) the Object Store, (*2*) the key, and (*3*) the value in the Object store.
image::os-tutorial-rtm-key-in-os.png[Key and value in the Object Store in Runtime Manager]
+
You see that the `myKey` key that you sent from `curl` appears in the Object Store with the value of `[binary value] BINARY`.
+
The `myKey` key that you sent from `curl` appears in the Object Store with the value of `[binary value] BINARY`.
The value is binary because Mule 4 wraps values in a Mule object, which causes
them to be visible only as binary in Anypoint Platform.

You have successfully tested your app using a `curl` command to store a key-value pair in the object store and viewed the results in Runtime Manager.

To verify that the value is correct, return to Design Center and add a new flow that retrieves the value.

> 

Reviewer: RESUME HERE

> === Create a Retrieve Flow to Retrieve the Value

Reviewer: Can we use this?
```suggestion
=== Create a Flow to Retrieve the Value
```

Reviewer: ```suggestion
=== Create a Flow to Retrieve the myKey Value
```
?

> To create a new flow to retrieve the value:

Reviewer: ```suggestion
To create a new flow to retrieve the 'myKey` value:
```

. In Design Center, in the *Project* area on the left, expand the *Flows* menu.
. Select *Rename* from the vertical ellipsis menu and rename your existing flow from `New Flow` to `Store`:
+
.The screenshot shows (*1*) the vertical ellipsis menu and (*2*) the *Rename* option.
+
image::os-tutorial-dc-rename-flow.png[Vertical ellipsis menu and Rename option]

. Click the plus sign to the right of *Flows*.
. In the wizard, click *Go straight to canvas*:
+
.The arrow shows the *Go straight to canvas* option.
image::os-tutorial-dc-goto-canvas.png[Go straight to canvas option]
+

> . Click the *Trigger* box and enter `H` to search for the HTTP Listener:

Reviewer: ```suggestion
. Select the *Trigger* box and enter `H` to search for the HTTP Listener:
```

Reviewer: Well, we select checkboxes. Is this "box" different?

Author: It's not a checkbox - it's actually box that you need to click to do something. I changed it to Click *Trigger*

+
.The arrow shows the HTTP Listener.  
image::os-tutorial-dc-http-search.png[Search for the HTTP Listener]
+
. Click *HTTP Listener*.

> . In the HTTP Listener window, set the value of *Path* to `/retrieve` and then click the *Close* button (X):

Reviewer: ```suggestion
. In the HTTP Listener window, set the value of *Path* to `/retrieve` and then click the close button (*X*):
```
If the window is labeled, that label should be formatted.

+
.The screenshot shows (*1*) the *Path* value and (*2*) the *Close* button (X).
image::os-tutorial-dc-http-retrieve-path.png[Path value and Close button]
+
Because the default settings for HTTP Listener are correct, you don't need to click *Edit*.

> . Rename this new flow to Retrieve.

Reviewer: ```suggestion
. Rename this new flow to `Retrieve`.
```

. Click the plus icon (+) to the right of the HTTP Listener trigger to add an Object Store Connector.
. In the *Select a component* window, enter `ob` to search for the ObjectStore Connector:
+
.The arrow shows the ObjectStore Connector.

> image::os-tutorial-dc-os-search.png[Search for the Object Store Connector]

Reviewer: GLOBAL:
Anypoint Connector for Object Store (Object Store Connector)
type of generic connector: Object Store connector
UI label/syntax: `ObjectStore Connector` or `ObjectStore` connector

+
. Click the connector, and then click the *Retrieve* operation.
. In the *Configuration* tab, specify `#[attributes.queryParams.key]` in the *Key* field:
+
.The arrow shows the *Key* field in the *Retrieve* configuration.
image::os-tutorial-dc-retrieve-values.png[Key field for the Retrieve operation]

. Click *Test* to run the app in a temporary location.

> . Click the copy icon to copy the deployment URL.

Reviewer: ```suggestion
. Click the copy icon to copy the deployment URL:
```

+
.The arrow shows the deployment URL copy icon.
image::os-tutorial-dc-copy-link.png[The copy deployment URL icon]

> . Run the following `curl` command from a command prompt to retrieve the key. 

Reviewer: ```suggestion
. From the command line, run the following `curl` command to retrieve the key, replacing  _DEPLOYMENT-URL_ with the deployment URL you copied from Design Center:
```

+

> Replace _DEPLOYMENT-URL_ with the deployment URL you copied from Design Center. 

Reviewer: see above

+
[source,console]
----
curl http://DEPLOYMENT-URL/retrieve?key=myKey;echo
----
+

> The output from this command lists the value for the `myKey`:

Reviewer: ```suggestion
The output from this command specifies the value for `myKey`:
```

Author: But it doesn't - it displays it.

Reviewer: Okay: displays works. I just don't suggest using "lists" for one thing.

+
[source,console]
----
"OSv2 Test"
----

You have successfully tested your app using a `curl` command to submit a REST API command to the Object Store API and to retrieve the results.

> Next, you create a Mule app in Anypoint Studio 7 and add an object store.

Reviewer: "an object store" vs. our branded "Object Store"? See Term Presentation List and let me know if it needs correcting?

> == Create and Configure a Mule 4 App with an Object Store in Anypoint Studio 7

Reviewer: Why "create" with "configure"; where is the "create" part?
Looks like "configure" and "deploy" to me!

Author: Already changed.

You can now use Anypoint Studio 7 to create a Mule 4 app similar to the app you created in Design Center.

First, you use Studio 7 to configure an app that includes an HTTP Listener trigger and an Object Store Connector.
Then, you deploy the app to CloudHub and use a `curl` command to send a key and value, and view the key in Runtime Manager.

To configure a Mule 4 app in Studio 7:

> . Start Studio, click *File* > *New* > *Mule Project*, and then name the project `osv2-demo`, and then click *Finish*.

Reviewer: ```suggestion
. Start Studio, click *File* > *New* > *Mule Project*, name the project `osv2-demo`, and then click *Finish*.
```
Thassa' lotta' actions for just one step....

Reviewer: GLOBALLY search for and affirm the proper presentation of all uses of the string "object store connector" and "objectstore connector" per previous edit comment. I'll leave them alone going forward.

. Install Object Store Connector:
.. Click the *Exchange* icon in the toolbar:
+
.The arrow shows the *Exchange* icon in the Studio 7 toolbar.
image::os-tutorial-studio7-exchange.png[Exchange icon in the Studio 7 toolbar]
.. Enter your Anypoint Platform username and password, if prompted.

> .. In Exchange, select *Provided by MuleSoft*, and then search for the ObjectStore Connector:

Reviewer: ```suggestion
.. In Exchange, select *Provided by MuleSoft* and then search for the ObjectStore Connector:
```

+
.The screenshot shows (*1*) the *Provided by MuleSoft* option and (*2*) the search field.
image::os-tutorial-exchange.png[Provided by MuleSoft option and search field in Exchange]
.. Select *ObjectStore Connector - Mule 4*.
+
Studio 7 supports Mule 4.
.. Click *Add to project*:
+
.The arrow shows the *Add to project* icon in Exchange.
image::os-tutorial-exchange-add-to-project.png[Add to project icon in Exchange]
.. At the confirmation prompt, click *Proceed* and then click *OK*.

> . In the *Mule Palette*, click *HTTP*, and then drag the *Listener* to the Studio canvas:

Reviewer: ```suggestion
. In *Mule Palette*, click *HTTP* and then drag the *Listener* to the Studio canvas:
```

+
.The screenshot shows (*1*) the HTTP Connector and (*2*) the Listener operation.
image::os-tutorial-studio7-palette.png[HTTP Connector and Listener operations in Mule Palette]

> . Click *ObjectStore*, and then drag the *Store* operation to the right of the Listener on the canvas.

Reviewer: ```suggestion
. Click *ObjectStore* and then drag the *Store* operation to the right of the Listener on the canvas.
```
...to the right of *Listener*....

+
.The arrow shows the *Store* operation in the Studio 7 canvas.
image::os-tutorial-studio7-flow.png[Store operation in the Studio 7 canvas]
. In the canvas, click *Listener*.
. In the Listener properties window, set *General > Path* to `/store`.
. Click the green plus icon (+) to the right of the *Connector configuration* field. 
. In the HTTP Listener *Global Element Properties* window, verify that *Host* is set to `0.0.0.0` and *Port* is set to `8081`, and then click *OK*.
+

> The Listener properties window now indicates that there are no errors:

Reviewer: ```suggestion
The Listener properties window indicates that there are no errors:
```

+
.The screenshot shows (*1*) the *Path* property and (*2*) the *Connector configuration* icon for the Listener.
image::os-tutorial-studio7-listener-props.png[Listener properties]
. In the canvas, click the *Store* operation.
. In the Store properties window, specify `#[payload.key]` in the *Key* field and `payload.value` in the *Value* field:
+
.The screenshot shows (*1*) the *Key* field and (*2*) the *Value* field.
image::os-tutorial-studio7-store-props.png[Key and Value fields for the Store operation]
+

> To use Object Store v2 with Mule 4 apps, do not select an object store global element in the *Object store* dropdown.

Reviewer: Per Term Presentation List, two entries:

"*Object Store* (branded internal service that does not require version)

*ObjectStore* commercial object database"

Reviewer: Lemme' know if I need to change/add to these.

Author: See slack

If you select the object store global element, the app uses the Mule runtime object store, rather than Object Store v2.
. Select the flow and, in the properties window, change the name to `Store`:
+
.The arrow shows the *Name* field in the Store flow properties window.
image::os-tutorial-studio7-store-flow-props.png[Name field in the Store flow properties window]

> . In the *Mule Palette* view, click *HTTP*, and then drag another *Listener* to the canvas below the first flow.

Reviewer: ```suggestion
. In the *Mule Palette* view, click *HTTP* and then drag another *Listener* to the canvas below the first flow.
```
You don't need the comma before "and then" in these steps because the steps have only two actions. GLOBAL

> . Click *ObjectStore*, and then drag the *Retrieve* operation to the right of the Listener in the second flow.

Reviewer: ```suggestion
. Click *ObjectStore* and then drag the *Retrieve* operation to the right of the Listener in the second flow.
```

Reviewer: GLOBALLY search for the string "listener" and verify the appropriate presentation:

listener, HTTP Listener, HTTP Listener component
vs.
syntax label in asterisks

Author: In Flow Designer, it's HTTP Listener.
In Studio 7, Listener is an operation within HTTP Connector, so "Listener" is correct.
In Studio 6, there is no Listener - you create HTTP flows for Store and Retrieve.

> . In the canvas, click the *Listener*.

Reviewer: ```suggestion
. In the canvas, click *Listener*.
```

. In the Listener properties window, set *General > Path* to `/retrieve`.

> . In the *Connector configuration* dropdown, select the same *HTTP_Listener_config* option that you set up for the Store Listener:

Reviewer: ```suggestion
. In the *Connector configuration* menu, select the same *HTTP_Listener_config* option that you set up for the Store Listener:
```

+
.The arrow shows the *Connector configuration* dropdown in the Listener properties window.
image::os-tutorial-studio7-retrieve-listener-props.png[Connector configuration dropdown in the Listener properties window]
. In the canvas, click the *Retrieve* operation.
. In the Retrieve properties window, specify `#[attributes.queryParams.key]` in the *Key* field:
+
.The arrow shows the *Key* field.
image::os-tutorial-studio7-retrieve-props.png[Key field for the Retrieve operation]
+
[IMPORTANT]

> With Mule 4, to use Object Store v2, do not select an object store global element in the *Object store* dropdown.

Reviewer: ```suggestion
When using Object Store in Mule 4, do not select an object store global element in the *Object store* menu.
```

. Select the second flow and, in the properties window, change the name to `Retrieve`.
+
When you are done, the flows should look like this:
+
.The screenshot shows (*1*) the Store flow and (*2*) the Retrieve flow.
image::os-tutorial-studio7-flows.png[Store and Retrieve flows in Studio]

. In *Package Explorer* in Studio, right-click `osv2-demo` and select *Anypoint Platform* > *Deploy to CloudHub*.
. At the prompt, select the environment to deploy the app to.
. In the Anypoint Platform *Deploying Application* window, click *Deploy Application*:
+
.The arrow shows the *Deploy Application* button in the *Deploying Application* page.
image::os-tutorial-deploy-app.png[Deploy Application button in the Deploying Application page]
. Click *Open in Browser* to display the app in Runtime Manager.
+
If CloudHub displays a permissions error, select a different environment to deploy to.

You have successfully used Studio 7 to configure an app with an object store.
Next, you test the connection by submitting REST API commands to the Object Store API and using Runtime Manager to view the results.

> === Test Your Studio 7 Application

Reviewer: BREAK

Reviewer: RESUME

include::_attributes.adoc[]
endif::[]

> Use this tutorial to get started with using Object Store v2 to store and view data.

Reviewer: ```suggestion
Use this tutorial to begin using Object Store v2 to store and view data.
```

> You can send any serializable data as key/value pairs to the object store.

Reviewer: ```suggestion
You can send any serializable data as key-value pairs to the object store.
```

> In this tutorial, you:

Reviewer: I can understand why you present the following info in the following way, but here are a couple of alternatives, if they aren't misleading....:
```suggestion
In this tutorial, you create two apps, one in Design Center and one in Anypoint Studio.
You then configure Anypoint Object Store Connector, send app data to the object store, and view that data in Runtime Manager.
```
OR
```suggestion
In this tutorial, you:
. Create apps in both Design Center and Anypoint Studio.
. Configure Anypoint Object Store Connector.
. Send data to the object store.
. View data in Runtime Manager.
```

Author: Neither captures the flow, though. And you configure the connector as part of app-creation. I used the same structure in the MQ tutorial, but there were additional steps, so maybe it made more sense?

. Create an app in Design Center, configure Anypoint Object Store Connector, and send data to the object store and view it in Runtime Manager.
. Create an app in Anypoint Studio, configure Anypoint Object Store Connector, and send data to the object store and view it in Runtime Manager.

> +

Reviewer: see above

Depending on your Mule version, you create your app in Studio 7 (Mule 4) or Studio 6 (Mule 3).

== Prerequisites

* Familiarity with Anypoint Studio, Anypoint Platform, and Runtime Manager
* An xref::anypoint.mulesoft.com[Anypoint Platform account]
+
If you don't have an Anypoint Platform login, click *Sign up* on the Anypoint Platform sign-in page.
* A Design environment configured in Anypoint Platform *Access Management > Environments*.
+
If you don't see *Access Management*, contact your site administrator.

> * Anypoint Studio installed on your system

Reviewer: I actually looked this up, and deeply apologize if I have edited this incorrectly in the past! It looks like we SHOULD use a colon before sublists and substeps:
```suggestion
* Anypoint Studio installed on your system:
```

+
--
** Studio 7 
+
See xref:studio::to-download-and-install-studio.adoc[Downloading and Installing Anypoint Studio].
** Studio 6 
+
See xref:6.5@studio::to-download-and-install-studio.adoc[Downloading and Installing Anypoint Studio].
--
* A https://curl.haxx.se[`curl`] command for testing 
+
This tutorial uses the HTTP Listener trigger to initiate Object Store Connector access and a `curl` command to send JSON data to your Mule app.
+
Instead of `curl`, you can use Postman or browser plugins.

== Create and Configure an App with an Object Store in Design Center

You use Design Center to configure an app that includes an HTTP Listener trigger and an Object Store Connector.
Then, you can use a `curl` command to send a key and value, and view the key in Runtime Manager.

=== Prerequisites

Disable password programs such as LastPass or Okta in your browser before using Design Center.
If enabled, the password program inserts passwords in fields in Design Center, which can cause the app to fail. 

> === Configure an App with Object Store in Design Center

Reviewer: If the parent topic is titled "Create and Configure an App with Object Store in Design Center," I expect to see a Create... topic before this Configure... topic.
And then I see that Configure is followed by peer Test and peer Create Retrieve?
I wonder if we should reeval the org. here.

Author: See below for new structure.

To configure an app in Design Center:

. In *Anypoint Platform*, click *Design Center* in the left navigation bar or the main Anypoint Platform screen.

> . Click the *Create new* button to create a new project.

Reviewer: Is "to create a new project" necessary?

Reviewer: c/b simply "Click *Create new*."

Reviewer: ...or Click *Create new* > *Create new application*:

+

> In the create window, select *Create new application*:

Reviewer: So...we have a topic Create and Configure.
Followed by subtopics Prerequisites and then Configure.
Configure is in AP, and you click Design Center > Create new > Create new application > New Mule Application > Create and then go to a wizard?
Just confirming the flow. These UI labels are so similar!

Reviewer: Is "the create window" actually "the *Create* window"?

Author: New structure:
== Prerequisites
== Create an App with an Object Store in Design Center
=== Test Your Design Center App
=== Create a Retrieve Flow to Retrieve the Value
== Create a Mule 4 App with an Object Store in Anypoint Studio 7
=== Test Your Studio 7 Application
=== Studio 7 XML
== Create a Mule 3 App with an Object Store in Anypoint Studio 6
=== Test Your Studio 6 Application
== See Also

+
.The arrow shows the *Create new application* option.
image::os-tutorial-dc-create-app.png[Create new application option]
+
. In the *New Mule Application* window, enter `osv2demo` for the project name and click *Create*.
. In the wizard, click *Go straight to canvas*:
+
.The arrow shows the *Go straight to canvas* option.
image::os-tutorial-dc-goto-canvas.png[Go straight to canvas]
+

> . Click the *Trigger* box and enter `H` to search for the HTTP Listener:

Reviewer: ```suggestion
. Select the *Trigger* box and enter `H` to search for the HTTP Listener:
```

+
.The arrow shows the HTTP Listener.  
image::os-tutorial-dc-http-search.png[Search for the HTTP Listener]
+
. Click *HTTP Listener*.

> . In the HTTP Listener window, set the value of *Path* to `/store` and then click the *Close* button (X):

Reviewer: label?
```suggestion
. In the *HTTP Listener* window, set the value of *Path* to `/store` and then click the close button (*X*):
```

+
.The screenshot shows (*1*) the *Path* value and (*2*) the *Close* button (X).
image::os-tutorial-dc-http-path.png[Path value and Close button]
+
Because the default settings for HTTP Listener are correct, you don't need to click *Edit*.

> . Click the plus icon (+) to the right of the HTTP Listener trigger:

Reviewer: ```suggestion
. Click the plus icon (*+*) to the right of the HTTP Listener trigger:
```

+
.The arrow shows the plus icon (+).
image::os-tutorial-dc-plus-icon.png[Plus icon]
+

> . In the *Select a component* window, enter `ob` to search for the ObjectStore Connector:

Reviewer: ```suggestion
. In the *Select a component* window, enter `ob` to search for the Object Store Connector:
```

+
.The arrow shows the ObjectStore Connector.
image::os-tutorial-dc-os-search.png[Search for the Object Store Connector]
+
. Click the connector, and then click the *Store* operation:
+
.The arrow shows the *Store* operation.
image::os-tutorial-dc-store-select.png[Store operation]
. In the *Configuration* tab, specify `#[payload.key]` in the *Key* field and `payload.value` in the *Value* field:
+
.The screenshot shows (*1*) the *Key* field and (*2*) the *Value* field.
image::os-tutorial-dc-store-values.png[Key and Value fields for the Store operation]
. Click *Test* to run the app in a temporary location.
+
The app starts running in a temporary environment:
+
.The arrow shows a message indicating the `osv2demo` app is running in a temporary environment.
image::os-tutorial-dc-running.png[The osv2demo app running in a temporary environment]

> You have successfully used Design Center to configure and deploy an app with an object store.

Reviewer: ...but if the app is running in a temp environment, is it really deployed?

Reviewer: Not too keen on saying "you have been successful" when we basically assume that our tasks work. 
I can see this for Citizen docs, but maybe not others.
How about "After you use Design Center to configure and deploy..., you can test the connection by...."

Reviewer: Also, if the culmination of this topic set is configuring and deploying, then the container topic should probably be Configure and Deploy rather than Create and Configure.

Author: See slack (and the MQ Tutorial).

> Next, you test the connection by submitting REST API commands to the Object Store API and using Runtime Manager to view the results.

Reviewer: So, if we are going to create a group of subtopics that "travels together," I suggest that their container explicitly list the tasks in order, not necessary by full title (and not links) but just to show the journey.

Author: I think we discussed this with the MQ Tutorial - having transitional summaries. Is this what you're talking about? I'm confused.

=== Test Your Design Center App

After you create your app in Design Center, you test your app using a `curl` command to submit a REST API command to the Object Store API and use Runtime Manager to view the results.

To test your app:

. Click the copy icon to copy the deployment URL:
+
.The arrow shows the deployment URL copy icon.
image::os-tutorial-dc-copy-link.png[The copy deployment URL icon]

> . Run the following `curl` command from a command prompt to send the key-value pair. 

Reviewer: ```suggestion
. From the command line, use `curl` to send the key-value pair, replacing _DEPLOYMENT-URL_ with...: 
```
"the" key-value pair? Which? Should we remind them?
Also, moved the next sentence to join the intro to the example (which ended with a colon...which should lead directly to the example).

+
Replace _DEPLOYMENT-URL_ with the deployment URL you copied from Design Center.
+
[source,json,linenums]
----

curl -X POST \
  "http://DEPLOYMENT-URL/store"
  -H "Content-Type: application/json" 
  -d '{ "key": "myKey", "value": "OSv2 Test" }' 

----
+
The output is an echo:
+
[source,console]
----
{ "key": "myKey", "value": "OSv2 Test" }
----
. Select *View in Runtime Manager* from the *...* menu to the right of the *Test* button:
+
.The arrow shows the *View in Runtime Manager* option in the *...* menu in Design Center.
image::os-tutorial-dc-view-rtm-option.png[View in Runtime Manager menu option]
. In *Runtime Manager*, click *Object Store* in the left navigation pane.
. Click the object store, the partition, and then the key:
+
.The screenshot shows (*1*) the Object Store, (*2*) the key, and (*3*) the value in the Object store.
image::os-tutorial-rtm-key-in-os.png[Key and value in the Object Store in Runtime Manager]
+

> You see that the myKey key that you sent from curl appears in the Object Store with the value of [binary value] BINARY.

Reviewer: ```suggestion
You see that the `myKey` key that you sent from `curl` appears in the Object Store with the value of `[binary value] BINARY`.
```

+

> *DO WE NEED THIS DETAIL?*

Reviewer: Be sure to comment this out.

+
The `myKey` key that you sent from `curl` appears in the Object Store with the value of `[binary value] BINARY`.
The value is binary because Mule 4 wraps values in a Mule object, which causes
them to be visible only as binary in Anypoint Platform.
Behind the scenes, Mule 4 executes binary serialization with the Mule internal serializer.

> The user interface cannot deserialize the object, so it shows that the value is now binary.

Reviewer: Even if you keep this text, I would omit anything that documents the back end and that the audience really doesn't need (or care) to know.

You have successfully tested your app using a `curl` command to store a key-value pair in the object store and viewed the results in Runtime Manager.

> To verify that the value is correct, return to Design Center and add a new flow that retrieves the value.

Reviewer: Ditto edit comments made at the completion of the previous task, re. "You have successfully" etc.

Reviewer: BREAK EDIT HERE

"OSv2 Test"
----

You have successfully tested your app using a `curl` command to store and retrieve a key-value pair in the object store and viewed the results in Runtime Manager.

=== Studio 7 XML

> You can use this XML to create the `osv2-demo` app in Studio 7.

Reviewer: ```suggestion
You can use XML to create the `osv2-demo` app in Studio 7 by clicking *Configuration XML* tab in Studio,  comparing your XML to the following, and making corrections as needed:
```

Click *Configuration XML* tab in Studio and compare your XML to the following.

> Make corrections if needed.

Reviewer: see above

    <http:listener doc:name="Listener" doc:id="805c84ec-35d4-4753-846b-49b108044df7" config-ref="HTTP_Listener_config" path="/store"/>
    <os:store doc:name="Store" doc:id="d4800501-0fa8-4472-a23e-5213f04155e8" key="#[payload.key]">
      <os:value ><![CDATA[#[payload.value]]]></os:value>
    </os:store>
  </flow>
  <flow name="Retrieve" doc:id="1db04ffc-76c1-48b4-ba33-57807a6404a8" >
    <http:listener doc:name="Listener" doc:id="faed2a0f-3005-4244-a6c6-77407b34d53a" path="/retrieve" config-ref="HTTP_Listener_config"/>
    <os:retrieve doc:name="Retrieve" doc:id="3f907cfc-59e5-49cd-879a-7b2675d488a5" key="#[attributes.queryParams.key]">
      <os:default-value ><![CDATA[#["TEST"]]]></os:default-value>
    </os:retrieve>
  </flow>
</mule>
----

== Create a Mule 3 App with an Object Store in Anypoint Studio 6

If you use Mule 3, you can use Anypoint Studio 6 to create a Mule 3 app similar to the app you created in Design Center.
In Studio 6, you can deploy the app to CloudHub and submit a `curl` command to test the connection. 

This section applies only to Anypoint Studio 6.
For Studio 7, see <<Create a Mule 4 App with an Object Store in Anypoint Studio 7>>.

> 

Reviewer: I wonder if we really need this, given that the title and both intro sentences specify Studio 6. Have customers been making this error?

Reviewer: lol...and the following two sentences also specify Studio 6.

> First, you use Studio 6 to configure an app that connects with an object store.

Reviewer: ```suggestion
First, you use Studio 6 to configure a Mule 3 app that connects with an object store:
```

> To configure a Mule 3 app in Studio 6:

Reviewer: See above

. Start Studio, click *File* > *New* > *Mule Project*, name the project `osv2-demo-studio-6`, and then click *Finish*.

. Install Object Store Connector:
.. Click the *Exchange* icon in the toolbar:
+
.The arrow shows the *Exchange* icon in the Studio 6 toolbar.
image::os-tutorial-studio6-exchange.png[Exchange icon in the Studio 6 toolbar]
.. Enter your Anypoint Platform username and password, if prompted.
.. In Exchange, select *Provided by MuleSoft*, and then search for the ObjectStore Connector:
+
.The screenshot shows (*1*) the *Provided by MuleSoft* option and (*2*) the search field.
image::os-tutorial-exchange.png[Provided by MuleSoft option and search field in Exchange]
.. Select *Object Store Connector - Mule 3*.
+

> Studio 6 supports Mule 3.

Reviewer: Again, do we have to reiterate this?

.. Click *Install*:
+
.The arrow shows the *Install* icon in Exchange.
image::os-tutorial-exchange-add-to-project-3.png[Install icon in Exchange]
.. At the prompt, check the items to install and click *Next* twice. 
.. Accept the license agreement and click *Finish*.
.. At the prompt, click *Restart Now* to finish installing the connector.

> . In the *Mule Palette*:

Reviewer: ```suggestion
. In *Mule Palette*:
```
OR
```suggestion
. In the *Mule Palette* view:
```

.. Drag *HTTP* to the Studio canvas:
+
.The arrow shows the *HTTP* Connector.
image::os-tutorial-studio6-palette.png[HTTP Connector in Mule Palette]

> .. Search for `string`, and then drag the *Object to String* component to the *Process* area to the right of HTTP on the canvas.

Reviewer: no need for comma in a progression of only two actions:
```suggestion
.. Search for `string` and then drag the *Object to String* component to the *Process* area to the right of HTTP on the canvas.
```
If HTTP is a label, format as such.

.. Search for `json`, and then drag the *JSON to Object* component to the canvas to the right of *Object to String*.
.. Search for `objects`, and then drag the *ObjectStore* connector to the canvas to the right of *JSON to Object*.

> .. Search for `payload`, and then drag the *Set Payload* component to the canvas to the right of *ObjectStore*.

Reviewer: ditto commas

+

> The flow should now look like this:

Reviewer: ```suggestion
The flow should look like this:
```

+
.The arrow shows the flow in the canvas.

> image::os-tutorial-studio6-flow-complete.png[Flow in Studio 6 Canvas]

Reviewer: ```suggestion
image::os-tutorial-studio6-flow-complete.png[Flow in Studio 6 canvas]
```

. In the canvas, click *HTTP*.
. In the HTTP properties window:
.. Set *Path* to `/store` and *Allowed Methods* to `POST`.
+
When you submit a curl command, you append the `/store` option to the URL.
The *Allowed Methods* setting ensures that the `curl` command can only POST information to the object store.
.. Click the green plus icon (+) to the right of *Connector Configuration*.

> .. In the HTTP Connector *Global Element Properties* window, verify that Host is set to 0.0.0.0 and Port is set to 8081, and then click *OK*.

Reviewer: ```suggestion
.. In the HTTP Connector *Global Element Properties* window, verify that *Host* is set to 0.0.0.0 and *Port* is set to 8081, and then click *OK*.
```
STET this comma because the first item in the list is a compound.

+
The HTTP properties window now indicates that there are no errors:
+
.The arrow shows the HTTP properties window.
image::os-tutorial-studio6-http-props.png[HTTP properties in Studio 6]

> . In the canvas, click *JSON to Object*, and then set *Return Class* to `java.util.Map`:

Reviewer: ```suggestion
. In the canvas, click *JSON to Object* and then set *Return Class* to `java.util.Map`:
```

+

> .The arrow shows the *Return Class* field in the JSON to Object properties window.

Reviewer: ```suggestion
.The arrow shows the *Return Class* field in the *JSON to Object* properties window.
```

image::os-tutorial-studio6-json-props.png[JSON to Object properties in Studio 6]
. In the canvas, click *ObjectStore*.

> . In the ObjectStore properties window:

Reviewer: ```suggestion
. In the *ObjectStore* properties window:
```
?

> .. Set the *Operation* to *Store* and set these store operation values:

Reviewer: ```suggestion
.. Set the *Operation* to `Store` and set these store operation values:
```

+
* Set *Key* to `#[payload.key]`.
* Set *Value Reference* to `#[payload.value]`.
.. Click the green plus icon (+) to the right of *Connector Configuration*.

> .. In the ObjectStore Connector *Global Element Properties* window, set *Partition* to `myPartition` and *Object Store Reference* to `_defaultUserObjectStore`, and then click *OK*.

Reviewer: ObjectStore Connector?
(Again, my head is spinning with all these permutations of object store...so I leave it in your capable hands. Even though I ask once in awhile....)

Author: See Slack. I evaluated them all and fixed some, but this is how it is in the UI...

+
.The screenshot shows (*1*) the *Partition* field and (*2*) the *Object Store Reference* field in the *Global Element Properties* window.
image::os-tutorial-studio6-os-global-props.png[Partion and Object Store Reference fields in the Global Element Properties window]
+
[NOTE]
The *Object Store Reference* field must be set to `_defaultUserObjectStore` for data to be stored in Object Store v2.
+
The ObjectStore properties window now indicates that there are no errors:

> +

Reviewer: BREAK for GitHub

Reviewer: RESUME

.The arrow shows the ObjectStore properties window.
image::os-tutorial-studio6-json-props.png[HTTP properties in Studio 6]
. In the canvas, click *Set Payload*.

> . In the Set Payload properties windows, set the *Value* field to `KEY: #[payload.key] AND VALUE: #[payload.value] WERE SAVED`,

Reviewer: Is Set Payload a label (*) and is there more than one window?

Reviewer: period instead of comma

Author: Set Payload is an operation.

+

> This message appears on the command line to indicate that the `curl` command values are received by the application.

Reviewer: colon

+
.The arrow shows the *Value* field in the Set Payload properties window.
image::os-tutorial-studio6-payload-props.png[Value field in the Set Payload properties window]
. Save your project.
. In *Package Explorer* in Studio, right-click `osv2-demo-studio-6` and select *Anypoint Platform* > *Deploy to Cloud*.
. At the prompt, select the environment to deploy the app to.

> . In the Anypoint Platform *Deploying Application* window, ensure that *Use Object Store v2* is selected, and then click *Deploy Application*:

Reviewer: ```suggestion
. In the Anypoint Platform *Deploying Application* window, ensure that *Use Object Store v2* is selected and then click *Deploy Application*:
```

+
.The arrow shows the *Deploy Application* button in the *Deploying Applications* page.
image::os-tutorial-deploy-app.png[Deploy Application button in the Deploying Applications page]
. Click *Open in Browser* to display the app in Runtime Manager.
+
If CloudHub displays a permissions error, select a different environment to deploy to.

You have successfully used Studio 6 to configure a Mule 3 app with an object store.
Next, you test the connection by submitting REST API commands to the Object Store API and using Runtime Manager to view the results.

=== Test Your Studio 6 Application

> After you create the app in Studio 6, you test your app using a `curl` command to submit a REST API command to the Object Store API and use Runtime Manager to view the results.

Reviewer: ```suggestion
After you create your app in Studio 6, you can test it by using a `curl` command to submit a REST API command to the Object Store API and use Runtime Manager to view the results.
```

To test your app:

. In *Runtime Manager*, click *Applications* and ensure that the `osv2-demo-studio-6` app is running.
. Click the status for the `osv2-demo-studio-6` app to display its details in the right pane.
. Copy the *App url* value:
+
.The arrow shows *App url* in the details pane in Runtime Manager.
image::os-tutorial-rtm-deploy-url-3.png[App url in Runtime Manager]
. Run the following `curl` command from a command prompt.
+

> Replace _APP-URL_ with the *App url* value that you copied from Runtime Manager. 

Reviewer: ditto prev edits to these two sentences

+
[source,console]
----
$ curl -X POST -H "Content-Type: application/json" -d '{ "key": "myStudio6Key", "value": "OSv2 Test from Studio 6" }' "http://APP-URL/store";echo
----
+
Each time you submit this command, change the key name so that each key is unique.
+

> The output is an echo:

Reviewer: ditto re. "echo"

+
[source,console]
----
KEY: myStudio6Key AND VALUE: OSv2 Test from Studio 6 WERE SAVED
----
+

> . In Runtime Manager, in the details pane, click *Manage application*.

Reviewer: ```suggestion
. In the Runtime Manager details pane, click *Manage application*.
```

Author: Not quite right. Will fix.

+
.The arrow shows *Manage application* in the details pane in Runtime Manager.
image::os-tutorial-rtm-manage-app-3.png[Manage application in the details pane]
. Click *Object Store* in the left navigation pane.
. Click the object store, the partition, and then the key:
+
.The screenshot shows (*1*) the *Object Store* tab, (*2*) the key, and (*3*) the value in the Object store.
image::os-tutorial-rtm-key-in-os-3.png[Object Store tab, key, and value in Runtime Manager]
. Hover over the *...* icon to display actions for each object:
+

> .The screenshot shows (*1*) the *Metadata* and (*2*) the *Delete Value* actions in the Object store.

Reviewer: Object store?

image::os-tutorial-rtm-os-actions-3.png[]
+
* To delete an object store, partition, key, or value, hover over the *...* icon and click the trash can icon.
* To display the key value and any metadata associated with the value, hover over the *...* icon and click the metadata icon.

== See Also

* xref:runtime-manager::deploying-to-cloudhub.adoc[Deploying to CloudHub]

> * xref:runtime-manager::runtime-manager-switch-env.adoc[Using Different Environments in Runtime Manager]

Reviewer: I've been bad about editing gerunds out of titles, but we need to start doing it going forward.

Author: Those are concepts. Should we change the link text for xrefs?
